examples/counter/scripts/abi-generator.ts

```diff
@@ -0,0 +1,117 @@
+// Clarinet extention 101
+//
+// This extension is introspecting the contracts of the project it's running from, 
+// and produce a Typescript data structure autocomplete friendly, that developers
+// can use in their frontend code.
+//
+// When running:
+// $ clarinet run --allow-write scripts/abi-generator.rs
+//
+// This script will write a file at the path artifacts/contracts.ts:
+//
+// export namespace CounterContract {
+//     export const ADDRESS = "ST000000000000000000002AMW42H";
+//     export const NAME = "counter";
+//     export enum Public {
+//         INCREMENT = "increment",
+//         DECREMENT = "decrement",
+//     }   
+//     export enum ReadOnly {
+//         READ = "read",
+//     }
+// }
+//
+// By importing this file in their frontend code, developers can use constants, instead 
+// of hard coding principals and strings:
+// 
+// await makeContractCall({
+//     contractAddress: CounterContract.ADDRESS,
+//     contractName: CounterContract.NAME,
+//     functionName: CounterContract.Public.INCREMENT,
+//     functionArgs: [CounterContract.Args.INCREMENT(namespace)], // or so
+//     ...
+
+import { Clarinet, Contract, Account, StacksNode } from 'https://deno.land/x/clarinet@v0.13.0/index.ts';
+
+Clarinet.run({
+    async fn(accounts: Map<string, Account>, contracts: Map<string, Contract>, node: StacksNode) {
+        let code = [];
+        code.push([
+            `// Code generated with the clarinet-abi-generator extension`,
+            `// Manual edits will be overwritten`,
+            ``,
+        ]);
+
+        for (let [contractId, contract] of contracts) {
+            let [address, name] = contractId.split(".");
+            code.push([
+                `export namespace ${kebabToCamel(capitalize(name))}Contract {`,
```

Comment:
I'm curious about the usage of typescript `namespace`, any specific reason for it? I don't think I've seen it used outside of .d.ts files before.

Reply:
I'm using a namespace to create some of scope, to avoid collisions when you have multiple contracts. How would you approach this?

Reply:
We can revisit the approach if it turns out being a disagreement for developers using this extension.
The goal of this PR is to demonstrate how to use extensions and how they can be used for automating tasks.